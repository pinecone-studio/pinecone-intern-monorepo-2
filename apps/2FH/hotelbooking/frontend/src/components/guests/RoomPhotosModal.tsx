/* eslint-disable  */
'use client';
import React, { useState } from 'react';
import { Dialog, DialogContent, DialogHeader, DialogTitle, DialogTrigger } from '@/components/ui/dialog';
import { Button } from '@/components/ui/button';
import { ChevronLeft, ChevronRight } from 'lucide-react';
import Image from 'next/image';

interface RoomPhotosModalProps {
  roomImages: string[];
  roomName: string;
}

const RoomPhotosModal = ({ roomImages, roomName }: RoomPhotosModalProps) => {
  const [open, setOpen] = useState(false);
  const [currentImageIndex, setCurrentImageIndex] = useState(0);

  const PLACEHOLDER_IMAGES = ['/placeholder-room-1.jpg', '/placeholder-room-2.jpg', '/placeholder-room-3.jpg', '/placeholder-room-4.jpg'];

  const getPlaceholderImages = () => PLACEHOLDER_IMAGES;

  // If no images provided, use placeholder images
  const images = roomImages && roomImages.length > 0 ? roomImages : getPlaceholderImages();

  const nextImage = () => {
    setCurrentImageIndex((prev) => (prev + 1) % images.length);
  };

  const prevImage = () => {
    setCurrentImageIndex((prev) => (prev - 1 + images.length) % images.length);
  };

  const goToImage = (index: number) => {
    setCurrentImageIndex(index);
  };

  return (
    <Dialog open={open} onOpenChange={setOpen}>
      <DialogTrigger asChild>
        <Button variant="link" className="text-blue-600 p-0 h-auto">
          <span className="flex items-center">
            <span className="mr-1">üëÅÔ∏è</span>
            View
          </span>
        </Button>
      </DialogTrigger>
      <DialogContent className="sm:max-w-4xl max-h-[90vh] overflow-hidden">
        <DialogHeader>
          <DialogTitle className="text-lg font-semibold">{roomName}</DialogTitle>
        </DialogHeader>

        <div className="space-y-4">
          {/* Main Image Display */}
          <div className="relative">
            <div className="relative h-96 w-full rounded-lg overflow-hidden bg-gray-100">
              <Image
                src={images[currentImageIndex]}
                alt={`${roomName} - Image ${currentImageIndex + 1}`}
                fill
                className="object-cover"
                onError={(e) => {
                  e.currentTarget.src = '/placeholder-room-1.jpg';
                }}
              />
            </div>

            {/* Navigation Arrows */}
            {images.length > 1 && (
              <>
                <Button
                  variant="ghost"
                  size="sm"
                  onClick={prevImage}
                  className="absolute left-2 top-1/2 transform -translate-y-1/2 bg-black/50 hover:bg-black/70 text-white"
                  aria-label="Previous image"
                >
                  <ChevronLeft className="h-4 w-4" />
                </Button>
                <Button variant="ghost" size="sm" onClick={nextImage} className="absolute right-2 top-1/2 transform -translate-y-1/2 bg-black/50 hover:bg-black/70 text-white" aria-label="Next image">
                  <ChevronRight className="h-4 w-4" />
                </Button>
              </>
            )}

            {/* Image Counter */}
            {images.length > 1 && (
              <div className="absolute bottom-2 left-1/2 transform -translate-x-1/2 bg-black/50 text-white px-2 py-1 rounded text-sm">
                {currentImageIndex + 1} / {images.length}
              </div>
            )}
          </div>

          {/* Thumbnail Navigation */}
          {images.length > 1 && (
            <div className="flex gap-2 overflow-x-auto pb-2">
              {images.map((image, index) => (
                <button
                  key={index}
                  onClick={() => goToImage(index)}
                  className={`relative h-16 w-24 rounded overflow-hidden flex-shrink-0 border-2 transition-all ${
                    index === currentImageIndex ? 'border-blue-500' : 'border-gray-200 hover:border-gray-300'
                  }`}
                >
                  <Image
                    src={image}
                    alt={`Thumbnail ${index + 1}`}
                    fill
                    className="object-cover"
                    onError={(e) => {
                      e.currentTarget.src = '/placeholder-room-1.jpg';
                    }}
                  />
                </button>
              ))}
            </div>
          )}
        </div>
      </DialogContent>
    </Dialog>
  );
};

export default RoomPhotosModal;
