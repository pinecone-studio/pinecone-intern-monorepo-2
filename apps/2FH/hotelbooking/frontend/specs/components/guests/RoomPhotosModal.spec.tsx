/* eslint-disable  */
import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import RoomPhotosModal from '@/components/guests/RoomPhotosModal';

// Mock dependencies
jest.mock('next/image', () => ({
  __esModule: true,
  default: ({ src, alt, className, onError }: any) => <img src={src} alt={alt} className={className} onError={onError} data-testid="modal-image" />,
}));

describe('RoomPhotosModal', () => {
  const mockRoomImages = ['https://example.com/image1.jpg', 'https://example.com/image2.jpg', 'https://example.com/image3.jpg'];

  const defaultProps = {
    roomImages: mockRoomImages,
    roomName: 'Deluxe Suite',
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders modal trigger button', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    expect(screen.getByText('View')).toBeInTheDocument();
  });

  it('opens modal when trigger button is clicked', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    expect(screen.getByText('Deluxe Suite')).toBeInTheDocument();
  });

  it('displays room name in modal header', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    expect(screen.getByText('Deluxe Suite')).toBeInTheDocument();
  });

  it('displays first image by default', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const images = screen.getAllByTestId('modal-image');
    const mainImage = images[0]; // Main image is always first
    expect(mainImage).toHaveAttribute('src', 'https://example.com/image1.jpg');
    expect(mainImage).toHaveAttribute('alt', 'Deluxe Suite - Image 1');
  });

  it('displays image counter when multiple images', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    expect(screen.getByText('1 / 3')).toBeInTheDocument();
  });

  it('displays navigation arrows when multiple images', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const prevButton = screen.getByRole('button', { name: /previous/i });
    const nextButton = screen.getByRole('button', { name: /next/i });

    expect(prevButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();
  });

  it('navigates to next image', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const nextButton = screen.getByRole('button', { name: /next/i });
    fireEvent.click(nextButton);

    const images = screen.getAllByTestId('modal-image');
    const mainImage = images[0]; // Main image is always first
    expect(mainImage).toHaveAttribute('src', 'https://example.com/image2.jpg');
    expect(screen.getByText('2 / 3')).toBeInTheDocument();
  });

  it('navigates to previous image', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const nextButton = screen.getByRole('button', { name: /next/i });
    fireEvent.click(nextButton);

    const prevButton = screen.getByRole('button', { name: /previous/i });
    fireEvent.click(prevButton);

    const images = screen.getAllByTestId('modal-image');
    const mainImage = images[0]; // Main image is always first
    expect(mainImage).toHaveAttribute('src', 'https://example.com/image1.jpg');
    expect(screen.getByText('1 / 3')).toBeInTheDocument();
  });

  it('wraps around to first image when navigating next from last image', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    // Navigate to last image
    const nextButton = screen.getByRole('button', { name: /next/i });
    fireEvent.click(nextButton);
    fireEvent.click(nextButton);

    // Navigate next from last image
    fireEvent.click(nextButton);

    const images = screen.getAllByTestId('modal-image');
    const mainImage = images[0]; // Main image is always first
    expect(mainImage).toHaveAttribute('src', 'https://example.com/image1.jpg');
    expect(screen.getByText('1 / 3')).toBeInTheDocument();
  });

  it('wraps around to last image when navigating previous from first image', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const prevButton = screen.getByRole('button', { name: /previous/i });
    fireEvent.click(prevButton);

    const images = screen.getAllByTestId('modal-image');
    const mainImage = images[0]; // Main image is always first
    expect(mainImage).toHaveAttribute('src', 'https://example.com/image3.jpg');
    expect(screen.getByText('3 / 3')).toBeInTheDocument();
  });

  it('displays thumbnail navigation when multiple images', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const thumbnails = screen.getAllByTestId('modal-image');
    expect(thumbnails).toHaveLength(4); // 1 main image + 3 thumbnails
  });

  it('navigates to image when thumbnail is clicked', async () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    // Find the thumbnail buttons by looking for buttons that contain images
    const thumbnailButtons = document.querySelectorAll('.flex.gap-2.overflow-x-auto.pb-2 button');
    const secondThumbnailButton = thumbnailButtons[1]; // Second thumbnail (index 1)
    fireEvent.click(secondThumbnailButton);

    // Wait for the state update to complete
    await waitFor(() => {
      const updatedImages = screen.getAllByTestId('modal-image');
      const mainImage = updatedImages[0]; // Main image (index 0)
      expect(mainImage).toHaveAttribute('src', 'https://example.com/image2.jpg');
    });

    expect(screen.getByText('2 / 3')).toBeInTheDocument();
  });

  it('highlights active thumbnail', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    // Find the thumbnail buttons by looking for buttons with thumbnail images
    const thumbnails = screen.getAllByTestId('modal-image');
    const firstThumbnailButton = thumbnails[1].closest('button'); // First thumbnail (index 1, after main image)
    expect(firstThumbnailButton).toHaveClass('border-blue-500');
  });

  it('handles single image without navigation', () => {
    const singleImage = ['https://example.com/single-image.jpg'];
    render(<RoomPhotosModal {...defaultProps} roomImages={singleImage} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    expect(screen.queryByText('1 / 1')).not.toBeInTheDocument();
    expect(screen.queryByRole('button', { name: /previous/i })).not.toBeInTheDocument();
    expect(screen.queryByRole('button', { name: /next/i })).not.toBeInTheDocument();
  });

  it('uses placeholder images when no images provided', () => {
    render(<RoomPhotosModal {...defaultProps} roomImages={[]} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const images = screen.getAllByTestId('modal-image');
    expect(images[0]).toHaveAttribute('src', '/placeholder-room-1.jpg');
  });

  it('uses placeholder images when roomImages is null', () => {
    render(<RoomPhotosModal {...defaultProps} roomImages={null as any} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const images = screen.getAllByTestId('modal-image');
    expect(images[0]).toHaveAttribute('src', '/placeholder-room-1.jpg');
  });

  it('uses placeholder images when roomImages is undefined', () => {
    render(<RoomPhotosModal {...defaultProps} roomImages={undefined as any} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const images = screen.getAllByTestId('modal-image');
    expect(images[0]).toHaveAttribute('src', '/placeholder-room-1.jpg');
  });

  it('handles image load error', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const images = screen.getAllByTestId('modal-image');
    const mainImage = images[0]; // Main image is the first one
    fireEvent.error(mainImage);

    expect(mainImage).toHaveAttribute('src', '/placeholder-room-1.jpg');
  });

  it('closes modal when open state changes', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    expect(screen.getByText('Deluxe Suite')).toBeInTheDocument();

    // Simulate modal close (this would normally be handled by the Dialog component)
    // In a real test, you might need to mock the Dialog component's onOpenChange
  });

  it('renders with proper modal structure', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const modalContent = document.querySelector('.sm\\:max-w-4xl.max-h-\\[90vh\\].overflow-hidden');
    expect(modalContent).toBeInTheDocument();
  });

  it('renders with proper image container', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const imageContainer = document.querySelector('.relative.h-96.w-full.rounded-lg.overflow-hidden');
    expect(imageContainer).toBeInTheDocument();
  });

  it('renders with proper thumbnail container', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    const thumbnailContainer = document.querySelector('.flex.gap-2.overflow-x-auto.pb-2');
    expect(thumbnailContainer).toBeInTheDocument();
  });

  it('handles room with very long name', () => {
    const longRoomName = 'This is a very long room name that might cause layout issues and should be handled properly';
    render(<RoomPhotosModal {...defaultProps} roomName={longRoomName} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    expect(screen.getByText(longRoomName)).toBeInTheDocument();
  });

  it('handles room with special characters in name', () => {
    const specialRoomName = 'Deluxe Suite & Resort (Premium) #101';
    render(<RoomPhotosModal {...defaultProps} roomName={specialRoomName} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    expect(screen.getByText(specialRoomName)).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<RoomPhotosModal {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with single image', () => {
    const singleImage = ['https://example.com/single-image.jpg'];
    const { container } = render(<RoomPhotosModal {...defaultProps} roomImages={singleImage} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with no images', () => {
    const { container } = render(<RoomPhotosModal {...defaultProps} roomImages={[]} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('handles thumbnail image error', () => {
    render(<RoomPhotosModal {...defaultProps} />);

    const triggerButton = screen.getByText('View');
    fireEvent.click(triggerButton);

    // Find thumbnail images and trigger error
    const thumbnailImages = screen.getAllByTestId('modal-image');
    const thumbnailImage = thumbnailImages[1]; // Second image (thumbnail)

    // Simulate image error
    fireEvent.error(thumbnailImage);

    // The onError callback should set placeholder image
    expect(thumbnailImage.src).toBe('http://localhost/placeholder-room-1.jpg');
  });
});
