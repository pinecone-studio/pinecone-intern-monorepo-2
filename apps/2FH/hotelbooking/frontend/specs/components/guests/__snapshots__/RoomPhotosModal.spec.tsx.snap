// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomPhotosModal matches snapshot 1`] = `
<button
  aria-controls="radix-:r28:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline text-blue-600 p-0 h-auto"
  data-state="closed"
  type="button"
>
  <span
    class="flex items-center"
  >
    <span
      class="mr-1"
    >
      👁️
    </span>
    View
  </span>
</button>
`;

exports[`RoomPhotosModal matches snapshot with no images 1`] = `
<button
  aria-controls="radix-:r2e:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline text-blue-600 p-0 h-auto"
  data-state="closed"
  type="button"
>
  <span
    class="flex items-center"
  >
    <span
      class="mr-1"
    >
      👁️
    </span>
    View
  </span>
</button>
`;

exports[`RoomPhotosModal matches snapshot with single image 1`] = `
<button
  aria-controls="radix-:r2b:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline text-blue-600 p-0 h-auto"
  data-state="closed"
  type="button"
>
  <span
    class="flex items-center"
  >
    <span
      class="mr-1"
    >
      👁️
    </span>
    View
  </span>
</button>
`;
